[UI] Redirect to previous filtered page after perform any action from index resource page

ernestWarwas opened the pull request:
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12 <!-- see the comment below -->          |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no <!-- don't forget to update the UPGRADE-*.md file --> |
| License         | MIT                                                          |

This pull request resolves the problem of redirecting to the not filtered page after eg. deleting the product or returning to page after editing some of resources by clicking cancel on update page.

<!--
 - Bug fixes must be submitted against the 1.10 or 1.11 branch(the lowest possible)
 - Features and deprecations must be submitted against the master branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

ernestWarwas committed:
[Behat] Added scenarios

[Behat] Add scenario

[Behat] new scenario added

ernestWarwas committed:
Logic implementation

Logic implementation

fix failing builds

failing builds fix

ernestWarwas committed:
RedirectHandler decorated

RedirectHandler decorated

check if page is opened

ernestWarwas committed:
removed unused contexts

ernestWarwas committed:
[Behat] Added new step to avoid problem with filtered page

[Behat] Added new step to avoid problem with filtered page

fixes

changed to new step in scenario

merge parameters and redirect to route directly

AdamKasp commented on src/Sylius/Bundle/AdminBundle/EventListener/AdminFilterSubscriber.php:
```diff
@@ -0,0 +1,91 @@
+final class AdminFilterSubscriber implements EventSubscriberInterface
+{
+    public function __construct(
+        private FilterStorageInterface $filterStorage,
```
can be in single line

AdamKasp commented on src/Sylius/Bundle/AdminBundle/EventListener/AdminFilterSubscriber.php:
```diff
@@ -0,0 +1,91 @@
+    ) {
+    }
+
+    public static function getSubscribedEvents()
```
```suggestion
    public static function getSubscribedEvents(): array
```

AdamKasp commented on src/Sylius/Bundle/AdminBundle/EventListener/AdminFilterSubscriber.php:
```diff
@@ -0,0 +1,91 @@
+            return;
+        }
+
+        $eventRequest = $event->getRequest();
```
minor thing but it can be before line 41 and you will be able to use this variable in the second `if`

GSadee commented on src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php:
```diff
@@ -342,6 +350,16 @@ public function iDeleteProduct(ProductInterface $product)
+    /**
+     * @When I delete the :product product on filtered page
+     */
+    public function iDeleteProductOnFilteredPage(ProductInterface $product)
```
Missing return type

GSadee commented on src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php:
```diff
@@ -435,6 +453,16 @@ public function iSaveMyChanges()
+    /**
+     * @When I cancel my changes
+     */
+    public function iCancelChanges()
```
Missing return type

GSadee commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,38 @@
+@admin_panel
+Feature: Redirecting to previous filtered page
```
Cover also a scenario for redirecting to the previous page, when there is more than 1 page and the user was on some other page than the first one

GSadee commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,38 @@
+        And the store has a product "FC Barcelona Home T-Shirt"
+        And I am logged in as an administrator
+
+    @ui
```
These scenarios should have `@no-api` tag I guess

GSadee commented on src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php:
```diff
@@ -1138,6 +1166,14 @@ public function theLastProductOnTheListShouldNotHaveName(): void
+    /**
+     * @Then I should be redirected to the previous filtered page with enabled filter
+     */
+    public function iShouldBeRedirectedToThePreviousFilteredPageWithFilter()
```
And missing return type

AdamKasp commented on src/Sylius/Bundle/AdminBundle/Storage/FilterStorageInterface.php:
```diff
@@ -0,0 +1,21 @@
+
+interface FilterStorageInterface
+{
+    public function set(array $filters): void;
+    public function all(): array;
+    public function hasFilters(): bool;
```
```suggestion
    public function set(array $filters): void;
    
    public function all(): array;
    
    public function hasFilters(): bool;
```

ernestWarwas committed:
[Behat] pagination scenario covered

[Behat] pagination scenario covered

[Behat] pagination scenario covered

[Behat] pagination scenario covered

[Behat] pagination scenario covered

[Behat] pagination scenario covered

check if route has additional parameters

remove unused declaration of service

remove unused declaration of service

reverted changes

renamed scenario

after review fixes

NoResponseMate commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        When I browse products
+        And I choose enabled filter
+        And I filter
+        And I go to 2 page
```
```suggestion
        And I go to the 2nd page
```

GSadee replied:
```suggestion
        And I go to the 2nd page
```
or
```suggestion
        And I go to the second page
```
?

NoResponseMate commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        And I am logged in as an administrator
+
+    @ui @no-api
+    Scenario: Redirecting to previous filtered page after delete product
```
```suggestion
    Scenario: Redirecting to previous filtered page after deleting a product
```

NoResponseMate commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        Then I should be redirected to the previous filtered page with enabled filter
+
+    @ui @no-api
+    Scenario: Redirecting to previous filtered page after cancelling editing product
```
```suggestion
    Scenario: Redirecting to previous filtered page after cancelling editing a product
```

NoResponseMate commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        Then I should be redirected to the previous filtered page with enabled filter
+
+    @ui @no-api
+    Scenario: Redirecting to previous filtered page with pagination after cancelling editing product
```
```suggestion
    Scenario: Redirecting to previous filtered page with pagination after cancelling editing a product
```

NoResponseMate commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+@admin_panel
+Feature: Redirecting to previous filtered page
```
I feel like 
```suggestion
Feature: Being redirected to previous filtered page
```
feels more natural. Same goes to all the scenarios.

NoResponseMate commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        And I go to 2 page
+        And I want to create a new simple product
+        And I cancel my changes
+        Then I should be redirected to the previous filtered page with enabled filter and 2 page
```
```suggestion
        Then I should be redirected to the 2nd page of only enabled products
```

NoResponseMate commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        Then I should be redirected to the previous filtered page with enabled filter and 2 page
+
+    @ui @javascript
+    Scenario: Redirecting to previous filtered page after cancelling creating new product
```
```suggestion
    Scenario: Redirecting to previous filtered page after cancelling creating a new product
```

NoResponseMate commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        Then I should be redirected to the previous filtered page with enabled filter and 2 page
+
+    @ui @javascript
+    Scenario: Redirecting to previous filtered page after cancelling editing product after creating new product
```
```suggestion
    Scenario: Redirecting to previous filtered page after cancelling editing product after creating a new product
```

NoResponseMate commented on src/Sylius/Behat/Context/Setup/ProductContext.php:
```diff
@@ -77,6 +77,18 @@ public function storeHasAProductPricedAt($productName, int $price = 100, Channel
     }
 
+    /**
+     * @Given the store has a :numberOfProducts products
```
```suggestion
     * @Given the store has :numberOfProducts products
```

NoResponseMate commented on src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php:
```diff
@@ -342,6 +350,16 @@ public function iDeleteProduct(ProductInterface $product)
         $this->indexPage->deleteResourceOnPage(['name' => $product->getName()]);
     }
 
+    /**
+     * @When I delete the :product product on filtered page
+     */
+    public function iDeleteProductOnFilteredPage(ProductInterface $product): void
+    {
+        $this->sharedStorage->set('product', $product);
+
+        $this->indexPage->deleteResourceOnPage(['name' => $product->getName()]);
+    }
+
```
Is this needed at all?

ernestWarwas replied:
Yes, because open again page resetting filters on page, and isOpen() function does not take into account filtered pages, and returning false even opened page is eg. product index. That's why I created this step.

NoResponseMate commented on src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php:
```diff
@@ -1138,6 +1174,23 @@ public function theLastProductOnTheListShouldNotHaveName(): void
     }
 
+    /**
+     * @Then I should be redirected to the previous filtered page with enabled filter
```
```suggestion
     * @Then I should be redirected to the previous page of only enabled products
```

NoResponseMate commented on src/Sylius/Bundle/UiBundle/Resources/views/Form/Buttons/_cancel.html.twig:
```diff
@@ -1 +1 @@
-<a href="{{ path|default(app.request.headers.get('referer')) }}" class="ui button">{{ 'sylius.ui.cancel'|trans }}</a>
+<a href="{{ path|default(app.request.headers.get('referer')) }}" class="ui button" id="sylius_cancel_changes_button">{{ 'sylius.ui.cancel'|trans }} </a>
```
```suggestion
<a href="{{ path|default(app.request.headers.get('referer')) }}" class="ui button" id="sylius_cancel_changes_button">{{ 'sylius.ui.cancel'|trans }}</a>
```

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/spec/EventListener/AdminFilterSubscriberSpec.php:
```diff
@@ -0,0 +1,210 @@
+        Request $request,
+        ParameterBag $attributes,
+        ParameterBag $query,
+        FilterStorageInterface $filterStorage
```
```suggestion
        FilterStorageInterface $filterStorage,
```
same with the rest functions here

NoResponseMate commented on src/Sylius/Behat/Page/Admin/Product/IndexPage.php:
```diff
@@ -74,13 +89,29 @@ public function checkLastProductHasDataAttribute(string $attributeName): bool
+        $element = $this->getElement('pagination_buttons');
+        $buttons = $element->find('css', '.active');
+
+        return (int) $buttons->getText();
```
```suggestion
        $buttons = $this->getElement('pagination_buttons');
        $activeButton = $buttons->find('css', '.active');

        return (int) $activeButton->getText();
```
or just inline the return
```suggestion
        $buttons = $this->getElement('pagination_buttons');

        return (int) $buttons->find('css', '.active')->getText();
```

GSadee replied:
Similar as the above, you can find the active one by proper element definition

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/EventListener/AdminFilterSubscriber.php:
```diff
@@ -0,0 +1,91 @@
+
+    private function isIndexResourceRoute(string $route): bool
+    {
+        return str_contains($route, 'index');
```
Why not
```suggestion
        return str_ends_with($route, 'index');
```
?

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/EventListener/AdminFilterSubscriber.php:
```diff
@@ -0,0 +1,91 @@
+
+    private function isSyliusRoute(string $route): bool
+    {
+        return str_starts_with($route, 'sylius');
```
I'm wondering whether it shouldn't be more lax here, to allow this functionality out of the box in custom plugin routes.

ernestWarwas replied:
yup, you're right. I Fixed it 💃

GSadee commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+    Background:
+        Given the store operates on a channel named "Poland"
+        And the store classifies its products as "Clothes"
+        And the store has a 15 products
```
```suggestion
        And the store has 15 products
```

GSadee commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        And I go to 2 page
+        And I want to modify the "Znicz Pruszków T-Shirt" product
+        And I cancel my changes
+        Then I should be redirected to the previous filtered page with enabled filter and 2 page
```
```suggestion
        Then I should be redirected to the 2nd page with enabled filter
```
or sth similar

GSadee commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        And I go to 2 page
+        And I create a new simple product "FC Barcelona Away T-Shirt" priced at "$20.00" with "Clothes" taxon in the "Poland" channel
+        And I cancel my changes
+        Then I should be redirected to the previous filtered page with enabled filter and 2 page
```
I'm not sure if this scenario gives any additional value as we already have `Redirecting to previous filtered page after cancelling editing product` above, I would be probably for removing this one

GSadee commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        And I cancel my changes
+        Then I should be redirected to the previous filtered page with enabled filter and 2 page
+
+    @ui @javascript
```
```suggestion
    @ui @javascript @no-api
```

GSadee commented on features/admin/redirecting_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,60 @@
+        And I cancel my changes
+        Then I should be redirected to the previous filtered page with enabled filter and 2 page
+
+    @ui @javascript
```
```suggestion
    @ui @javascript @no-api
```

GSadee commented on src/Sylius/Behat/Page/Admin/Product/IndexPage.php:
```diff
@@ -64,6 +70,15 @@ public function showProductPage(string $productName): void
+        /** @var NodeElement $element */
+        $element = $this->getElement('pagination_buttons');
+        $buttons = $element->findAll('css', '.item');
+        $button = $buttons[$page];
```
Probably it can be refactored to find the proper button using the `getElement` method as we know the number of the page. The element should look similar to: `'pagination_button' => '.sylius-grid-nav__pagination a.item:contains("%page%")',`

GSadee commented on src/Sylius/Bundle/AdminBundle/EventListener/AdminFilterSubscriber.php:
```diff
@@ -0,0 +1,91 @@
+
+        if (!$this->isSyliusRoute($originalRoute) ||
+            !$this->isAdminSection($requestAttributes->get('_sylius', []))
+        ) {
```
```suggestion
        if (
            !$this->isSyliusRoute($originalRoute) ||
            !$this->isAdminSection($requestAttributes->get('_sylius', []))
        ) {
```

GSadee commented on src/Sylius/Bundle/AdminBundle/spec/Storage/FilterStorageSpec.php:
```diff
@@ -0,0 +1,73 @@
+        $this->shouldImplement(FilterStorageInterface::class);
+    }
+
+    function it_sets_a_filters_in_a_session(SessionInterface $session): void
```
```suggestion
    function it_sets_filters_in_a_session(SessionInterface $session): void
```

ernestWarwas committed:
[Behat] Renamed scenarios and feature file

[Behat] Renamed scenarios and feature file

Removed checking if route is sylius route

NoResponseMate commented on features/admin/being_redirected_to_previous_filtered_page.feature:
```diff
@@ -0,0 +1,50 @@
+@admin_panel
+Feature: Being redirected to previous filtered page
+    In order to have proper filtered page
+    As an Administrator
+    I want to be able to redirect to previous filtered page after any action from index
```
```suggestion
    In order to return to a properly filtered page
    As an Administrator
    I want to be redirected to a previously filtered page after taking any action on index
```

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/EventListener/AdminFilterSubscriber.php:
```diff
@@ -0,0 +1,84 @@
+            return;
+        }
+
+        if (!$this->isIndexResourceRoute($originalRoute)) {
```
Moving this condition above the other 2 would most likely speed things up.

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/Resources/config/services.xml:
```diff
@@ -60,6 +60,16 @@
+            <tag name="twig.extension" />
+        </service>
+
+        <service id="Sylius\Bundle\AdminBundle\Storage\FilterStorage">
```
Can this service be private?

ernestWarwas replied:
I didn’t check that all services here are public. Good catch! Thanks 😃

ernestWarwas committed:
Privatize filter storage service

GSadee commented on src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php:
```diff
@@ -389,6 +407,14 @@ public function iWantToModifyAProduct(ProductInterface $product): void
     }
 
+    /**
+     * @When /^I go to ([^"]+)(nd) page$/
```
```suggestion
     * @When /^I go to the (\d)(?:st|nd|rd|th) page$/
```

Zales0123 commented on src/Sylius/Behat/Page/Admin/Product/IndexPage.php:
```diff
@@ -74,13 +85,27 @@ public function checkLastProductHasDataAttribute(string $attributeName): bool
+        return $this->getElement('enabled_filter')->getValue() === 'true';
+    }
+
+    public function checkPageNumber(): int
```
`check` assumes assertion, it's rather `get` 🖖

Zales0123 commented on src/Sylius/Bundle/AdminBundle/Resources/config/services/controller.xml:
```diff
@@ -76,5 +76,13 @@
+        <service id="Sylius\Bundle\AdminBundle\Controller\RedirectHandler"
+                 decorates="sylius.resource_controller.redirect_handler"
+                 public="false"
+        >
```
```suggestion
        <service
            id="Sylius\Bundle\AdminBundle\Controller\RedirectHandler"
            decorates="sylius.resource_controller.redirect_handler"
            public="false"
        >
```
🚀

Zales0123 commented on src/Sylius/Bundle/AdminBundle/Twig/FilterExtension.php:
```diff
@@ -0,0 +1,66 @@
+use Twig\Extension\AbstractExtension;
+use Twig\TwigFunction;
+
+final class FilterExtension extends AbstractExtension
```
`RedirectPathExtension`?

Zales0123 replied:
Additionally, I know it's a twig extension and they are usually well tested by Behat scenarios, but this one contains a lot of logic, so it deserves at least a unit test

Zales0123 commented on src/Sylius/Bundle/UiBundle/Resources/views/Form/Buttons/_cancel.html.twig:
```diff
@@ -1 +1 @@
-<a href="{{ path|default(app.request.headers.get('referer')) }}" class="ui button">{{ 'sylius.ui.cancel'|trans }}</a>
+<a href="{{ path|default(app.request.headers.get('referer')) }}" class="ui button" id="sylius_cancel_changes_button">{{ 'sylius.ui.cancel'|trans }}</a>
```
We add this id only for Behat purposes, we should use test attribute instead

Zales0123 commented on src/Sylius/Bundle/AdminBundle/EventListener/AdminFilterSubscriber.php:
```diff
@@ -0,0 +1,84 @@
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+final class AdminFilterSubscriber implements EventSubscriberInterface
```
I wonder, do we need a subscriber for that? The only moment when we need to set the filters is when we press the "Filter" button (which means - we access the `indexAction` with some `criteria` in the query parameters). The moment when we need to remove them is when we access the `indexAction` without criteria in the query parameters. Is it possible to just override/decorate the `indexAction` by default to take it into account? 🤔 We would be able to skip the subscriber then 🚀

Zales0123 replied:
Second thought :D Probably it would be hard to enhance an `indexAction` for all resources other than changing it in the ResourceBundle itself 💃

Zales0123 replied:
Final thought 🚀 Let's leave it as it is right now and try to resolve it later her or on ResourceBundle itself 🖖

Zales0123 committed:
Naming and code style changes

Zales0123 committed:
Add test for twig extension to ensure proper path generation

NoResponseMate commented on src/Sylius/Behat/Page/Admin/Product/IndexPageInterface.php:
```diff
@@ -25,9 +25,17 @@ public function showProductPage(string $productName): void;
 
     public function checkLastProductHasDataAttribute(string $attributeName): bool;
+
+    public function getPageNumber(): int;
```
```suggestion
    public function getActivePageNumber(): int;
```
or
```suggestion
    public function getCurrentPageNumber(): int;
```
to give a bit more context

TheMilek commented on src/Sylius/Behat/Resources/config/suites/ui/admin/panel.yml:
```diff
@@ -0,0 +1,27 @@
+                - sylius.behat.context.transform.product
+                - sylius.behat.context.transform.taxon
+
+                - sylius.behat.context.setup.admin_user
+                - sylius.behat.context.setup.channel
+                - sylius.behat.context.setup.product
+                - sylius.behat.context.setup.admin_security
+                - sylius.behat.context.setup.taxonomy
```
```suggestion
                - sylius.behat.context.setup.admin_security
                - sylius.behat.context.setup.admin_user
                - sylius.behat.context.setup.channel
                - sylius.behat.context.setup.product
                - sylius.behat.context.setup.taxonomy
```
alphabetical order :v:

GSadee commented:
Thank you, Ernest and Mateusz! :1st_place_medal:
